### PR TITLE
Refactor WebRTC

### PR DESCRIPTION
**Pull request by njooma**
Refactor WebRTC code and implement DialOptions

### CODE REVIEW

File: example/viam_example_app/lib/main.dart [hunk 1]
```diff
@@ -89,7 +89,9 @@ class _MyHomePageState extends State<MyHomePage> {
       _loading = true;
     });
     final robotFut = RobotClient.atAddress(
-        '<URL>', 443, RobotClientOptions.withSecret('<SECRET>'));
+      '<URL>',
+      RobotClientOptions.withLocationSecret('<SECRET>'),
```
Comment (clintpurser) [inline example/viam_example_app/lib/main.dart:93]: ✨

File: lib/src/media/image.dart [hunk 2]
```diff
@@ -118,8 +118,6 @@ class _ViamRGBADecoder extends img.Decoder {
 
   @override
   img.Image? decodeFrame(int frame) {
-    Uint8List imageData;
```
Comment (clintpurser) [inline lib/src/media/image.dart:121]: [q] why remove this here? is it a _driveby_ that you noticed was unused?

File: lib/src/media/stream/client.dart [hunk 3]
```diff
@@ -44,14 +43,13 @@ class StreamClient {
     return _streamController.stream;
   }
 
-  @override
   Future<void> add(String name) async {
     await _client.addStream(AddStreamRequest(name: _getValidSDPTrackName(name)));
   }
 
-  @override
```
Comment (clintpurser) [inline lib/src/media/stream/client.dart:52]: good catch on these 👍

File: lib/src/robot/client.dart [hunk 4]
```diff
@@ -1,36 +1,35 @@
...
 
+import '../domain/web_rtc/web_rtc_client/web_rtc_client.dart';
+import '../gen/common/v1/common.pb.dart';
+import '../gen/robot/v1/robot.pbgrpc.dart';
 import '../media/stream/client.dart';
+import '../resource/base.dart';
+import '../resource/manager.dart';
+import '../resource/registry.dart';
+import '../rpc/dial.dart';
+import '../viam_sdk.dart';
```
Comment (clintpurser) [inline lib/src/robot/client.dart:11]: I think we can add a lint rule to prefer relative imports. 👀  but maybe we don't want that because we'll want to do package imports in our example code. something to think about!

File: lib/src/rpc/dial.dart [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+      try {
+        final candidateProto = ICECandidate(
+          candidate: candidate.candidate,
+          sdpMid: candidate.sdpMid,
+          sdpmLineIndex: candidate.sdpMLineIndex,
+        );
+        await signalingClient.callUpdate(CallUpdateRequest(uuid: uuid, candidate: candidateProto));
+      } catch (error, st) {
+        //TODO: Add error handling
+        _logger.e('Update ICECandidate error', error, st);
```
Comment (clintpurser) [inline lib/src/rpc/dial.dart:172]: Thanks for setting up logger! 🙏

File: lib/src/rpc/dial.dart [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+      if (!isResponseStreamInitialized) {
+        _logger.e('Updating before initialized');
+        throw Exception('Updating before initialized');
+      }
+      if (response.uuid != uuid) {
+        _logger.e('UUID mismatch in update. Got ${response.uuid}, want $uuid');
+        throw Exception('UUID mismatch in update. Got ${response.uuid}, want $uuid');
+      }
+
+      // await Future.delayed(const Duration(seconds: 1));
```
Comment (clintpurser) [inline lib/src/rpc/dial.dart:256]: [minor] remove?

File: lib/src/rpc/dial.dart [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+  String? externalAuthToEntity;
+
+  /// Allows a pre-authenticated client to dial with
+  /// an authorization header. Direct dial will have the access token
+  /// appended to the "Authorization: Bearer" header. WebRTC dial will
+  /// append it to the signaling server communication
+  ///
+  /// If enabled, other auth options have no affect. Eg. [authEntity], [credentials],
+  /// [externalAuthAddress], [externalAuthToEntity]
+  String? accessToken;
```
Comment (clintpurser) [inline lib/src/rpc/dial.dart:39]: was there a particular reason you made these Strings`?` nullable?

IMO It adds a lot of boiler plate below to check if they're null and then if they're empty.

could we make them non nullable and initialize them to '' to skip those null checks?

```suggestion
  String authEntity = '';

  /// Used to authenticate the request.
  Credentials? credentials;

  /// Options to control how WebRTC is utilized in a dial attempt
  DialWebRtcOptions? webRtcOptions;

  /// The address to perform external auth at.
  String externalAuthAddress  = '';

  /// The entity to authenticate for after externally authenticating.
  String externalAuthToEntity  = '';

  /// Allows a pre-authenticated client to dial with
  /// an authorization header. Direct dial will have the access token
  /// appended to the "Authorization: Bearer" header. WebRTC dial will
  /// append it to the signaling server communication
  ///
  /// If enabled, other auth options have no affect. Eg. [authEntity], [credentials],
  /// [externalAuthAddress], [externalAuthToEntity]
  String accessToken  = '';
```

File: lib/src/robot/client.dart [hunk 8]
```diff
@@ -1,36 +1,35 @@
...
 
 class RobotClient {
   late ClientChannelBase channel;
   late RobotServiceClient _client;
   List<ResourceName> resourceNames = [];
   ResourceManager _manager = ResourceManager();
-  Map<String, StreamClient> _streams = {};
-  late Viam viam;
+  final Map<String, StreamClient> _streams = {};
+  Viam? viam;
```
Comment (clintpurser) [inline lib/src/robot/client.dart:26]: [minor] I think this is unused, remove?

a viam is referenced on lines 38-41, but that name is overloaded and the new one will be used and not this one.

File: lib/src/rpc/dial.dart [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+  Credentials? signalingCredentials;
+
+  /// Allows a pre-authenticated client to dial with an authorization header to the signaling server.
+  /// This skips the Authenticate() request to the signaling server or external auth but does not skip the
+  /// AuthenticateTo() request to retrieve the credentials at the external auth endpoint.
+  ///
+  /// If enabled, other auth options have no affect. Eg. [signalingAuthEntity], [signalingCredentials],
+  /// [signalingExternalAuthAddress], [signalingExternalAuthToEntity]
+  String? signalingAccessToken;
+}
```
Comment (clintpurser) [inline lib/src/rpc/dial.dart:84]: [minor] same suggestion as above, possibly remove nullable

**Review by clintpurser — CHANGES_REQUESTED** [review]
Overall looks great! Just a few nits and minor suggestions.

File: lib/src/media/image.dart (hunk 2 above)
Comment (njooma) [reply lib/src/media/image.dart:121]: Yup, unused

File: lib/src/rpc/dial.dart (hunk 7 above)
Comment (njooma) [reply lib/src/rpc/dial.dart:40]: Decided against doing this (because I liked the nil coalescing later on for convenience), so I instead made an extension on `String?` that will check for both null and empty. Makes things the cleanest :)

File: lib/src/rpc/dial.dart (hunk 9 above)
Comment (njooma) [reply lib/src/rpc/dial.dart:85]: Same as above

File: lib/src/rpc/dial.dart [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+      ..id = 1
+      ..negotiated = true
+      ..ordered = true,
+  );
+
+  String? uuid;
+  final didConnect = Completer();
+  final didSetRemoteDesc = Completer();
+
+  // If tickleICE is enabled, set onIceCandidate handler
```
Comment (benjirewis) [inline lib/src/rpc/dial.dart:154]:
```suggestion
  // If trickleICE is enabled, set onIceCandidate handler
```

**Review by benjirewis — COMMENTED** [review]
Cool! 🧑‍🔧

File: lib/src/rpc/dial.dart [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+
+      try {
+        final candidateProto = ICECandidate(
+          candidate: candidate.candidate,
+          sdpMid: candidate.sdpMid,
+          sdpmLineIndex: candidate.sdpMLineIndex,
+        );
+        await signalingClient.callUpdate(CallUpdateRequest(uuid: uuid, candidate: candidateProto));
+      } catch (error, st) {
+        //TODO: Add error handling
```
[the remark this answers is not in the document]
Comment (njooma) [reply lib/src/rpc/dial.dart:172]: That todo is incorrect. We're not supposed to rethrow errors on call updates.

File: lib/src/rpc/dial.dart [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+      ..negotiated = true
+      ..ordered = true,
+  );
+
+  String? uuid;
+  final didConnect = Completer();
+  final didSetRemoteDesc = Completer();
+
+  // If tickleICE is enabled, set onIceCandidate handler
+  if (!(options.webRtcOptions?.disableTrickleIce ?? config.disableTrickle)) {
```
[the remark this answers is not in the document]
Comment (njooma) [reply lib/src/rpc/dial.dart:155]: `config.disableTrickle` was received by the server as an optional webrtc configuration. It shouldn't override `webrtcoptions.disable...` but it can be used as a backup in case the value the user passed in is empty

**Review by clintpurser — APPROVED** [review]
LGTM

File: lib/src/robot/client.dart [hunk 13]
```diff
@@ -1,53 +1,50 @@
...
   RobotClient._();
 
-  static Future<RobotClient> atAddress(String url, int port, RobotClientOptions options) async {
+  static Future<RobotClient> atAddress(String url, RobotClientOptions options) async {
     var client = RobotClient._();
-    client.viam = Viam.instance();
-    await client.viam
-        .connect(url: url, port: port, secure: !options.insecure, disableWebRtc: options.disableWebRtc, payload: options.locationSecret);
-    client.channel = client.viam.channel;
+    client.channel = await dial(url, options.dialOptions);
```
[the remark this answers is not in the document]
Comment (njooma) [reply lib/src/robot/client.dart:35]: It didn't before, but it does now :)

**Review by benjirewis — APPROVED** [review]
Nice, connecting local addresses will be helpful if we implement modules in dart.

File: lib/src/rpc/dial.dart [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+  });
+  await peerConnection.setLocalDescription(offer);
+
+  final sdpJsonString = _convertSDPtoJsonString(await peerConnection.getLocalDescription());
+  final encodedSdp = _encodeSDPJsonStringToBase64String(sdpJsonString);
+  try {
+    callStream = signalingClient
+        .call(CallRequest(sdp: encodedSdp, disableTrickle: options.webRtcOptions?.disableTrickleIce ?? config.disableTrickle));
+  } catch (error, st) {
+    //TODO: Add error handling
```
[the remark this answers is not in the document]
Comment (njooma) [reply lib/src/rpc/dial.dart:243]: This is a different TODO that I didn't catch earlier. But only the `CallUpdate` shouldn't rethrow, since it's not really required for everything to function

File: lib/src/rpc/dial.dart [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+  }
+  final iceServers = config.additionalIceServers
+      .map((e) => {
+            'urls': e.urls,
+            'credential': e.credential,
+            'username': e.username,
+          })
+      .toList()
+    ..add({
+      'urls': 'stun:global.stun.twilio.com:3478?transport=udp',
```
[the remark this answers is not in the document]
Comment (njooma) [reply lib/src/rpc/dial.dart:149]: I copied this over from what our friends over at DOR had done. Just tested it, seems like we don't need it
